cpp/include/cuml/experimental/fil/decision_forest.hpp

```diff
@@ -0,0 +1,466 @@
+      categorical_storage_.has_value() ? categorical_storage_->data() : static_cast<categorical_storage_type*>(nullptr)
+    );
+    switch(nodes_.device().index()) {
+      case 0:
+        fil::detail::infer(
+          obj(),
+          get_postprocessor(),
+          output.data(),
+          input.data(),
+          index_type(input.size() / num_feature_),
+          num_feature_,
+          num_outputs_,
+          has_categorical_nodes_,
+          vector_output_data,
+          categorical_storage_data,
+          specified_rows_per_block_iter,
+          std::get<0>(nodes_.device()),
+          stream
+        );
+        break;
+      case 1:
+        fil::detail::infer(
+          obj(),
+          get_postprocessor(),
+          output.data(),
+          input.data(),
+          index_type(input.size() / num_feature_),
+          num_feature_,
+          num_outputs_,
+          has_categorical_nodes_,
+          vector_output_data,
+          categorical_storage_data,
+          specified_rows_per_block_iter,
+          std::get<1>(nodes_.device()),
+          stream
+        );
+        break;
+    }
+  }
+
+ private:
+  /** The nodes for all trees in the forest */
+  raft_proto::buffer<node_type> nodes_;
+  /** The index of the root node for each tree in the forest */
+  raft_proto::buffer<index_type> root_node_indexes_;
+  /** Buffer of outputs for all leaves in vector-leaf models */
+  std::optional<raft_proto::buffer<io_type>> vector_output_;
+  /** Buffer of outputs for all leaves in vector-leaf models */
+  std::optional<raft_proto::buffer<categorical_storage_type>> categorical_storage_;
+
+  // Metadata
+  index_type num_feature_;
+  index_type num_outputs_;
+  index_type leaf_size_;
+  bool has_categorical_nodes_ = false;
+  // Postprocessing constants
+  row_op row_postproc_;
+  element_op elem_postproc_;
+  io_type average_factor_;
+  io_type bias_;
+  io_type postproc_constant_;
+
+  auto obj() const {
+    return forest_type{
+      nodes_.data(),
+      root_node_indexes_.data(),
+      static_cast<index_type>(root_node_indexes_.size())
+    };
+  }
+
+  auto get_postprocessor() const {
+    return postprocessor_type {
+      row_postproc_,
+      elem_postproc_,
+      average_factor_,
+      bias_,
+      postproc_constant_
+    };
+  }
+
+  auto leaf_size() const { return leaf_size_; }
+};
+
+namespace detail {
+/**
+ * A convenience wrapper to simplify template instantiation of
+ * decision_forest
+ *
+ * This template takes the large range of available template parameters
+ * and reduces them to just three standard choices.
+ *
+ * @tparam layout The in-memory layout of nodes in this forest
+ * @tparam double_precision Whether this model should use double-precision
+ * for floating-point evaluation and 64-bit integers for indexes
+ * @tparam large_trees Whether this forest expects more than 2**(16 -3) - 1 =
+ * 8191 features or contains nodes whose child is offset more than 2**16 - 1 = 65535 nodes away.
```

Comment:
I started to go through the process of doing this, but I would prefer to derive them as constexpr values from our specialization types rather than hardcoding them as "magic values." That in turn requires a bit more refactoring than I'd prefer to do at this stage, so if it's all right, I'll leave it for a follow-up.